### ISSUE TITLE
The version of cuda?

### ISSUE THREAD

**Issue by hanjiang2001** [sunoh-kim/pps#1]
Hello, I'd like to reproduce your excellent code. However, I'm a bit confused by the environment specified in the ReadMe ("cuda 11.6+pytorch 2.0.1"). On the website https://pytorch.org/get-started/previous-versions/#v201, I only see CUDA versions 11.7 and 11.8 listed. Could this be a mistake in describing the environment you are using?

**Comment by sunoh-kim** [comment]
Thank you for your interest. As I remember, I used cuda 11.6 with torch 2.0.1. However, I think torch version is more important for reproduction. So I recommend using torch 2.0.1 and cuda 11.7 or 11.8.

**Comment by hanjiang2001** [comment]
Thank you for your reply, and I would also like to know if this is a necessary modification regarding the difference between your Refact version and the original version, and what is the reason for the change? Because I observed that this resulted in a slight decrease in the results.

**Comment by sunoh-kim** [comment]
There was replacemnet of my desktop (change in my workplace :) ) and some code refactoring. However, the biggest reason for the performance change is due to the environmental change that occurred when I replaced my desktop. The perfrmance not only decreased but also increased at some metrics. For example, at Rank5@0.5 on the ActivityNet Captions dataset, the performance improved significantly from 71.32 to 74.85. I provide both models before the change and after the change and their hyper-parameters.

**Comment by hanjiang2001** [comment]
Sorry to bother you again, but after matching with your environment (which is probably only different from your cuda version), I think that my reproduction effect on the characters dataset is not satisfactory. Specifically, the result obtained by using config-react.json is "2024-04-11 23:41:52684- | Rank1, mIoU 0.4458 | Rank1.", IoU@0.3 0.6786 | Rank1, IoU@0.5 0.4956 | Rank1, IoU@0.7 0.2562 | Rank5, mIoU 0.6863 | Rank5, IoU@0.3 0.9829 | Rank5, IoU@0.5 0.8610 | Rank5, IoU@0.7 0.5279 |. The result obtained using config.json is "2024-04-12 05:00:30002- | Rank1, mIoU 0.4452 | Rank1, IoU@0.3 0.6792 | Rank1, IoU@0.5 0.4991 | Rank1, IoU@0.7 0.2426 | Rank5, mIoU 0.6826 | Rank5, IoU@0.3 0.9823 | Rank5, IoU@0.5 0.8521 | Rank5, IoU@0.7 0.5063 |.
I have provided the environment dependencies I am using as follows. Could you please share the dependencies you are using to help me approach the results reported in your paper(or the refact result)? If I receive your help, I would greatly appreciate it.

dependencies:
  - _libgcc_mutex=0.1=main
  - _openmp_mutex=5.1=1_gnu
  - bzip2=1.0.8=h5eee18b_5
  - ca-certificates=2024.3.11=h06a4308_0
  - ld_impl_linux-64=2.38=h1181459_1
  - libffi=3.4.4=h6a678d5_0
  - libgcc-ng=11.2.0=h1234567_1
  - libgomp=11.2.0=h1234567_1
  - libstdcxx-ng=11.2.0=h1234567_1
  - libuuid=1.41.5=h5eee18b_0
  - ncurses=6.4=h6a678d5_0
  - openssl=1.1.1w=h7f8727e_0
  - pip=23.3.1=py310h06a4308_0
  - python=3.10.8=h7a1cb2a_1
  - readline=8.2=h5eee18b_0
  - setuptools=68.2.2=py310h06a4308_0
  - sqlite=3.41.2=h5eee18b_0
  - tk=8.6.12=h1ccaba5_0
  - tzdata=2024a=h04d1e81_0
  - wheel=0.41.2=py310h06a4308_0
  - xz=5.4.6=h5eee18b_0
  - zlib=1.2.13=h5eee18b_0
  - pip:
    - antlr4-python3-runtime==4.8
    - bitarray==2.9.2
    - certifi==2024.2.2
    - cffi==1.16.0
    - charset-normalizer==3.3.2
    - click==8.1.7
    - cmake==3.29.1
    - colorama==0.4.6
    - cython==3.0.10
    - fairseq==0.12.2
    - filelock==3.13.4
    - h5py==3.8.0
    - hydra-core==1.0.7
    - idna==3.6
    - jinja2==3.1.3
    - joblib==1.4.0
    - lit==18.1.2
    - lxml==5.2.1
    - markupsafe==2.1.5
    - mpmath==1.3.0
    - networkx==3.3
    - nltk==3.8.1
    - numpy==1.26.4
    - nvidia-cublas-cu11==11.10.3.66
    - nvidia-cuda-cupti-cu11==11.7.101
    - nvidia-cuda-nvrtc-cu11==11.7.99
    - nvidia-cuda-runtime-cu11==11.7.99
    - nvidia-cudnn-cu11==8.5.0.96
    - nvidia-cufft-cu11==10.9.0.58
    - nvidia-curand-cu11==10.2.10.91
    - nvidia-cusolver-cu11==11.4.0.1
    - nvidia-cusparse-cu11==11.7.4.91
    - nvidia-nccl-cu11==2.14.3
    - nvidia-nvtx-cu11==11.7.91
    - omegaconf==2.0.6
    - pillow==10.3.0
    - portalocker==2.8.2
    - pycparser==2.22
    - pyyaml==6.0.1
    - regex==2023.12.25
    - requests==2.31.0
    - sacrebleu==2.4.1
    - sympy==1.12
    - tabulate==0.9.0
    - torch==2.0.1
    - torchaudio==2.0.2
    - torchvision==0.15.2
    - tqdm==4.66.2
    - triton==2.0.0
    - typing-extensions==4.11.0
    - urllib3==2.2.1

**Comment by hanjiang2001** [comment]
By the way, through the model file you provided, I can easily obtain the same results as the report in your paper. This is really a great job. Thank you for sharing!

**Comment by sunoh-kim** [comment]
> Sorry to bother you again, but after matching with your environment (which is probably only different from your cuda version), I think that my reproduction effect on the characters dataset is not satisfactory. Specifically, the result obtained by using config-react.json is "2024-04-11 23:41:52684- | Rank1, mIoU 0.4458 | Rank1.", IoU@0.3 0.6786 | Rank1, IoU@0.5 0.4956 | Rank1, IoU@0.7 0.2562 | Rank5, mIoU 0.6863 | Rank5, IoU@0.3 0.9829 | Rank5, IoU@0.5 0.8610 | Rank5, IoU@0.7 0.5279 |. The result obtained using config.json is "2024-04-12 05:00:30002- | Rank1, mIoU 0.4452 | Rank1, IoU@0.3 0.6792 | Rank1, IoU@0.5 0.4991 | Rank1, IoU@0.7 0.2426 | Rank5, mIoU 0.6826 | Rank5, IoU@0.3 0.9823 | Rank5, IoU@0.5 0.8521 | Rank5, IoU@0.7 0.5063 |. I have provided the environment dependencies I am using as follows. Could you please share the dependencies you are using to help me approach the results reported in your paper(or the refact result)? If I receive your help, I would greatly appreciate it.
> 
> dependencies:
> 
> * _libgcc_mutex=0.1=main
> * _openmp_mutex=5.1=1_gnu
> * bzip2=1.0.8=h5eee18b_5
> * ca-certificates=2024.3.11=h06a4308_0
> * ld_impl_linux-64=2.38=h1181459_1
> * libffi=3.4.4=h6a678d5_0
> * libgcc-ng=11.2.0=h1234567_1
> * libgomp=11.2.0=h1234567_1
> * libstdcxx-ng=11.2.0=h1234567_1
> * libuuid=1.41.5=h5eee18b_0
> * ncurses=6.4=h6a678d5_0
> * openssl=1.1.1w=h7f8727e_0
> * pip=23.3.1=py310h06a4308_0
> * python=3.10.8=h7a1cb2a_1
> * readline=8.2=h5eee18b_0
> * setuptools=68.2.2=py310h06a4308_0
> * sqlite=3.41.2=h5eee18b_0
> * tk=8.6.12=h1ccaba5_0
> * tzdata=2024a=h04d1e81_0
> * wheel=0.41.2=py310h06a4308_0
> * xz=5.4.6=h5eee18b_0
> * zlib=1.2.13=h5eee18b_0
> * pip:
>   
>   * antlr4-python3-runtime==4.8
>   * bitarray==2.9.2
>   * certifi==2024.2.2
>   * cffi==1.16.0
>   * charset-normalizer==3.3.2
>   * click==8.1.7
>   * cmake==3.29.1
>   * colorama==0.4.6
>   * cython==3.0.10
>   * fairseq==0.12.2
>   * filelock==3.13.4
>   * h5py==3.8.0
>   * hydra-core==1.0.7
>   * idna==3.6
>   * jinja2==3.1.3
>   * joblib==1.4.0
>   * lit==18.1.2
>   * lxml==5.2.1
>   * markupsafe==2.1.5
>   * mpmath==1.3.0
>   * networkx==3.3
>   * nltk==3.8.1
>   * numpy==1.26.4
>   * nvidia-cublas-cu11==11.10.3.66
>   * nvidia-cuda-cupti-cu11==11.7.101
>   * nvidia-cuda-nvrtc-cu11==11.7.99
>   * nvidia-cuda-runtime-cu11==11.7.99
>   * nvidia-cudnn-cu11==8.5.0.96
>   * nvidia-cufft-cu11==10.9.0.58
>   * nvidia-curand-cu11==10.2.10.91
>   * nvidia-cusolver-cu11==11.4.0.1
>   * nvidia-cusparse-cu11==11.7.4.91
>   * nvidia-nccl-cu11==2.14.3
>   * nvidia-nvtx-cu11==11.7.91
>   * omegaconf==2.0.6
>   * pillow==10.3.0
>   * portalocker==2.8.2
>   * pycparser==2.22
>   * pyyaml==6.0.1
>   * regex==2023.12.25
>   * requests==2.31.0
>   * sacrebleu==2.4.1
>   * sympy==1.12
>   * tabulate==0.9.0
>   * torch==2.0.1
>   * torchaudio==2.0.2
>   * torchvision==0.15.2
>   * tqdm==4.66.2
>   * triton==2.0.0
>   * typing-extensions==4.11.0
>   * urllib3==2.2.1


If you have any questions please feel free to ask :)

Are you checking the performance on the last epoch? It would be better to find the epoch with the best performance rather than the last epoch. Of course, there are many evaluation metrics, so it would be good to check which point is the best epoch by visualizing the results through wandb.

Due to changes in my workplace, the GPU was changed from A100 to RTX4090, resulting in slight differences in performance. This performance change is similar to what happened with our base code, which is CPL (https://github.com/minghangz/cpl). My guess is that this is due to the transformer code (transformer.py), but I am not sure.

I recommend checking the performance changes by adjusting the hyperparameters "alpha_ivc", "alpha_pull", "alpha_inter_push", and "alpha_intra_push" that control our Gaussian mixture proposal generation. In my case, when I changed these hyperparameters, I can reproduce the results at the different environment (refactored version).

I do not know the exact versions of all dependencies because my development environment has changed now, but I remember that except for cuda and torch versions, there were no factors that affected performance at that time. I think matching the versions in the readme would be sufficient.

**Comment by hanjiang2001** [comment]
Thank you so much, I think I will try to adjust the hyperparameters.
